--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -553,6 +_,13 @@
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
             return level.getBlockState(pos).is(Blocks.LAVA) && level.getBlockState(pos.above()).isPathfindable(PathComputationType.LAND);
         }
+
+        // cookie start - Async target finding
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.Strider;
+        }
+        // cookie end - Async target finding
     }
 
     static class StriderPathNavigation extends GroundPathNavigation {
@@ -560,9 +_,25 @@
             super(strider, level);
         }
 
+        // cookie start - async path processing
+        private static final io.cookiemc.cookie.entity.pathfinding.NodeEvaluatorGenerator nodeEvaluatorGenerator = (io.cookiemc.cookie.entity.pathfinding.NodeEvaluatorFeatures nodeEvaluatorFeatures) -> {
+            WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // cookie end
+
         @Override
         protected PathFinder createPathFinder(int maxVisitedNodes) {
             this.nodeEvaluator = new WalkNodeEvaluator();
+            // cookie start - async path processing
+            if (io.cookiemc.cookie.Config.INSTANCE.entities.pathfinding.enableThreadedPathfinding) {
+                return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+            }
+            // cookie end
             return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
         }
 
