--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -407,7 +_,7 @@
         @Override
         public void tick() {
             BlockPos boundOrigin = Vex.this.getBoundOrigin();
-            if (boundOrigin == null) {
+            if (boundOrigin == null || !io.cookiemc.cookie.region.ServerRegions.isTickThreadFor(Vex.this.level(), boundOrigin)) { // cookie - Threaded Regions
                 boundOrigin = Vex.this.blockPosition();
             }
 
