--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1,5 +_,8 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.misc.NearbyPlayers;
+import java.util.concurrent.atomic.AtomicReference;
+import io.canvasmc.canvas.Config;
 import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
@@ -103,8 +_,6 @@
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Unit;
-import net.minecraft.util.profiling.Profiler;
-import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.world.Container;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.InteractionHand;
@@ -188,7 +_,9 @@
 
 public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patches.chunk_system.player.ChunkSystemServerPlayer { // Paper - rewrite chunk system
     private static final Logger LOGGER = LogUtils.getLogger();
-    public long lastSave = MinecraftServer.currentTick; // Paper - Incremental chunk and player saving
+    // Canvas start - Threaded Regions
+    public static final long LAST_SAVE_ABSENT = Long.MIN_VALUE;
+    public long lastSave = LAST_SAVE_ABSENT; // Paper - Incremental chunk and player saving
     private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_XZ = 32;
     private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_Y = 10;
     private static final int FLY_STAT_RECORDING_SPEED = 25;
@@ -257,7 +_,6 @@
     private Vec3 lastKnownClientMovement = Vec3.ZERO;
     private Input lastClientInput = Input.EMPTY;
     private final Set<ThrownEnderpearl> enderPearls = new HashSet<>();
-    public final AtomicReference<NearbyPlayers> npr = new AtomicReference<>(); // Canvas - Threaded Regions
     public final ContainerSynchronizer containerSynchronizer = new ContainerSynchronizer() {
         private final LoadingCache<TypedDataComponent<?>, Integer> cache = CacheBuilder.newBuilder()
             .maximumSize(256L)
@@ -423,7 +_,7 @@
     private boolean tpsBar = false; // Purpur - Implement TPSBar
     private boolean compassBar = false; // Purpur - Add compass command
     private boolean ramBar = false; // Purpur - Implement rambar commands
+    public volatile boolean requestingTimeResync = false; // Canvas - Threaded Dimensions
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
@@ -492,7 +_,7 @@
             long l1 = l * l;
             int i = l1 > 2147483647L ? Integer.MAX_VALUE : (int)l1;
             int coprime = this.getCoprime(i);
-            int randomInt = RandomSource.create().nextInt(i);
+            int randomInt = level.random.nextInt(i); // Canvas - reduce RandomSource instances
 
             for (int i1 = 0; i1 < i; i1++) {
                 int i2 = (randomInt + coprime * i1) % i;
@@ -771,12 +_,21 @@
 
         Entity camera = this.getCamera();
         if (camera != this) {
-            if (camera.isAlive()) {
+            if (camera.canBeSpectated()) { // Canvas - Threaded Regions - replace removed check
+                if (io.canvasmc.canvas.region.ServerRegions.isTickThreadFor(camera) && !camera.isRemoved()) { // Canvas - Threaded Regions
                 this.absSnapTo(camera.getX(), camera.getY(), camera.getZ(), camera.getYRot(), camera.getXRot());
                 this.serverLevel().getChunkSource().move(this);
                 if (this.wantsToStopRiding()) {
                     this.setCamera(this);
                 }
+                // Canvas start - Threaded Regions
+                } else {
+                    Entity realCamera = camera.getBukkitEntity().getHandleRaw();
+                    if (realCamera != camera) {
+                        this.setCamera(this);
+                        this.setCamera(realCamera);
+                    } else {
+                        this.teleportToCameraOffRegion();
+                    }
+                }
+                // Canvas end
             } else {
                 this.setCamera(this);
             }
@@ -801,23 +_,18 @@
         }
         // Purpur end - Ridables
     }
+    // Canvas start - Threaded Regions
 
     private void teleportToCameraOffRegion() {
-        Entity cameraFinal = this.getCamera();
-        // use the task scheduler, as we don't know where the caller is invoking from
+        final Entity cameraFinal = this.getCamera();
         if (this != cameraFinal) {
-            this.getBukkitEntity().taskScheduler.schedule((final ServerPlayer newPlayer) -> {
-                io.canvasmc.canvas.region.ServerRegions.teleport(
-                    newPlayer, false, cameraFinal, null, null,
-                    org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE, null,
-                    (final ServerPlayer newerPlayer) -> {
-                        return newerPlayer.camera == cameraFinal;
-                    }
-                );
-            }, null, 1L);
-        } // else: do not bother teleporting to self
+            io.canvasmc.canvas.region.ServerRegions.teleportAsync(this, cameraFinal.getBukkitEntity().getLocation(), org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE)
+                .exceptionally(ex -> {
+                    LOGGER.warn("Failed to teleport player {} to spectate {}", this.getScoreboardName(), cameraFinal.getScoreboardName(), ex);
+                    return false;
+                });
+        }
     }
-    // Canvas end
 
     private void updatePlayerAttributes() {
         AttributeInstance attribute = this.getAttribute(Attributes.BLOCK_INTERACTION_RANGE);
@@ -842,15 +_,22 @@
     public void doTick() {
         try {
             if (valid && !this.isSpectator() || !this.touchingUnloadedChunk()) { // Paper - don't tick dead players that are not in the world currently (pending respawn)
-                super.tick();
+                if (!this.touchingUnloadedChunk()) super.tick(); // Canvas - ensure we dont touch any unloaded chunk
             }
 
+            // Canvas start - only tick items in hand
+            if (Config.INSTANCE.entities.onlyTickItemsInHand) {
+                this.synchronizeSpecialItemUpdates(this.getMainHandItem());
+                this.synchronizeSpecialItemUpdates(this.getOffhandItem());
+            } else {
+            // Canvas end
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack item = this.getInventory().getItem(i);
                 if (!item.isEmpty()) {
                     this.synchronizeSpecialItemUpdates(item);
                 }
             }
+            } // Canvas - only tick items in hand
 
             if (this.getHealth() != this.lastSentHealth
                 || this.lastSentFood != this.foodData.getFoodLevel()
@@ -921,7 +_,7 @@
             }
             // CraftBukkit end
         } catch (Throwable var4) {
+            var4.printStackTrace(); // Canvas - print please.
             CrashReport crashReport = CrashReport.forThrowable(var4, "Ticking player");
             CrashReportCategory crashReportCategory = crashReport.addCategory("Player being ticked");
             this.fillCrashReportCategory(crashReportCategory);
@@ -1044,7 +_,7 @@
     // Paper end - Expand PlayerDeathEvent API
     @Override
     public void die(DamageSource cause) {
+        processingDeath = true; // Canvas - item entity spread
         // this.gameEvent(GameEvent.ENTITY_DIE); // Paper - move below event cancellation check
         boolean _boolean = this.serverLevel().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES); final boolean showDeathMessage = _boolean; // Paper - OBFHELPER
         // CraftBukkit start - fire PlayerDeathEvent
@@ -1160,7 +_,7 @@
         this.getCombatTracker().recheckStatus();
         this.setLastDeathLocation(Optional.of(GlobalPos.of(this.level().dimension(), this.blockPosition())));
         this.setClientLoaded(false);
+        processingDeath = false; // Canvas
     }
 
     private void tellNeutralMobsThatIDied() {
@@ -1422,18 +_,13 @@
                 this.unsetRemoved();
                 */
                 // CraftBukkit end
-                ProfilerFiller profilerFiller = Profiler.get();
-                profilerFiller.push("moving");
                 if (level != null && resourceKey == net.minecraft.world.level.dimension.LevelStem.OVERWORLD && level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER) { // CraftBukkit - empty to fall through to null to event
                     this.enteredNetherPosition = this.position();
                 }
 
-                profilerFiller.pop();
-                profilerFiller.push("placing");
                 // CraftBukkit start
                 this.isChangingDimension = true; // CraftBukkit - Set teleport invulnerability only if player changing worlds
                 LevelData worlddata = level.getLevelData();
@@ -1451,7 +_,6 @@
                 this.connection.internalTeleport(PositionMoveRotation.of(teleportTransition), teleportTransition.relatives()); // CraftBukkit - use internal teleport without event
                 this.connection.resetPosition();
                 level.addDuringTeleport(this);
-                profilerFiller.pop();
                 this.triggerDimensionChangeTriggers(serverLevel);
                 this.stopUsingItem();
                 this.connection.send(new ClientboundPlayerAbilitiesPacket(this.getAbilities()));
@@ -1534,7 +_,6 @@
         super.take(entity, quantity);
         this.containerMenu.broadcastChanges();
     }
-    // Canvas start - cache eligible players for despawn checks
 
     @Override
     public boolean isAlive() {
@@ -2166,19 +_,18 @@
 
     @Override
     public void crit(Entity entityHit) {
-        this.serverLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(entityHit, 4));
+        // Canvas start - broadcast crit animations as the entity being critted
+        var level = this.serverLevel();
+        level.getChunkSource().broadcastAndSend(Config.INSTANCE.fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entityHit : this, new ClientboundAnimatePacket(entityHit, 4));
+        // Canvas end
     }
 
     @Override
     public void magicCrit(Entity entityHit) {
-        this.serverLevel().getChunkSource().broadcastAndSend(this, new ClientboundAnimatePacket(entityHit, 5));
+        // Canvas start - broadcast crit animations as the entity being critted
+        var level = this.serverLevel();
+        level.getChunkSource().broadcastAndSend(Config.INSTANCE.fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entityHit : this, new ClientboundAnimatePacket(entityHit, 5));
+        // Canvas end
     }
 
     @Override
@@ -2202,7 +_,7 @@
             this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, gameMode.getId()));
             if (gameMode == GameType.SPECTATOR) {
                 this.removeEntitiesOnShoulder();
+                this.stopSleeping(); // Canvas - MC-119417
                 this.stopRiding();
                 EnchantmentHelper.stopLocationBasedEffects(this);
             } else {
@@ -2478,10 +_,11 @@
     }
 
     public void setCamera(@Nullable Entity entityToSpectate) {
+        // Canvas start - Threaded Regions
+        if (entityToSpectate != null && (entityToSpectate != this && !entityToSpectate.canBeSpectated())) {
+            return;
+        }
+        // Canvas end
         Entity camera = this.getCamera();
         this.camera = (Entity)(entityToSpectate == null ? this : entityToSpectate);
         if (camera != this.camera) {
@@ -2500,18 +_,16 @@
                 }
             }
             // Paper end - Add PlayerStartSpectatingEntityEvent and PlayerStopSpectatingEntity
-            if (this.camera.level() instanceof ServerLevel serverLevel) {
-                this.teleportTo(serverLevel, this.camera.getX(), this.camera.getY(), this.camera.getZ(), Set.of(), this.getYRot(), this.getXRot(), false, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
-            }
-
-            if (entityToSpectate != null) {
-                this.serverLevel().getChunkSource().move(this);
-            }
-
+            // Canvas - Threaded Regions - move down && not needed
+
+            // Canvas start - Threaded Regions - handle camera setting better
+            if (this.camera == this
+                || (io.canvasmc.canvas.region.ServerRegions.isTickThreadFor(this.camera) && this.camera.moonrise$getTrackedEntity() != null
+                && this.camera.moonrise$getTrackedEntity().seenBy.contains(this.connection))) {
             this.connection.send(new ClientboundSetCameraPacket(this.camera));
-            this.connection.resetPosition();
+            }
+            // this.connection.resetPosition();
+            this.teleportToCameraOffRegion();
+            // Canvas end
         }
     }
 
