--- a/net/minecraft/world/entity/projectile/LlamaSpit.java
+++ b/net/minecraft/world/entity/projectile/LlamaSpit.java
@@ -47,6 +_,11 @@
     @Override
     public void tick() {
         super.tick();
+        // cookie start - Threaded Regions
+        if (!io.cookiemc.cookie.region.ServerRegions.isTickThreadFor(this.level(), this.position(), this.getDeltaMovement(), 1)) {
+            return;
+        }
+        // cookie end
         Vec3 deltaMovement = this.getDeltaMovement();
         HitResult hitResultOnMoveVector = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
         this.preHitTargetOrDeflectSelf(hitResultOnMoveVector); // CraftBukkit - projectile hit event
