--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -44,6 +_,11 @@
 
     @Override
     public void tick() {
+        // cookie start - Threaded Regions
+        if (!io.cookiemc.cookie.region.ServerRegions.isTickThreadFor(this.level(), this.position(), this.getDeltaMovement(), 1)) {
+            return;
+        }
+        // cookie end
         this.handleFirstTickBubbleColumn();
         this.applyGravity();
         this.applyInertia();
