--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -3,6 +_,8 @@
 import java.util.Objects;
 import java.util.UUID;
 import javax.annotation.Nullable;
+import io.cookiemc.cookie.Config;
+import io.cookiemc.cookie.region.ServerRegions;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.UUIDUtil;
 import net.minecraft.nbt.CompoundTag;
@@ -160,8 +_,26 @@
     }
     // Paper end - EAR 2
 
+    // cookie start
+    private void lagCompensation() {
+        ServerRegions.WorldTickData tickData = ServerRegions.getTickData(this.level().level()); // cookie - Threaded Regions
+        if (!Config.INSTANCE.lagCompensation.enabled || !Config.INSTANCE.lagCompensation.pickupAcceleration) return;
+        if ((this).level().isClientSide()) return;
+
+        if (pickupDelay == 0) return;
+
+        if (pickupDelay - tickData.tpsCalculator.applicableMissedTicks() <= 0) { // cookie - Threaded Regions
+            pickupDelay = 0;
+            return;
+        }
+
+        pickupDelay = pickupDelay - tickData.tpsCalculator.applicableMissedTicks(); // cookie - Threaded Regions
+    }
+
     @Override
     public void tick() {
+        lagCompensation();
+    // cookie end
         if (this.getItem().isEmpty()) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
@@ -216,7 +_,8 @@
                 || Mth.floor(this.yo) != Mth.floor(this.getY())
                 || Mth.floor(this.zo) != Mth.floor(this.getZ());
             int i = flag ? 2 : 40;
-            if (this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable()) {
+            ItemStack item = this.getItem(); // cookie - short cut count check
+            if (item.getCount() < item.getMaxStackSize() && (this.tickCount % i == 0 && !this.level().isClientSide && this.isMergable())) { // cookie - short cut count check
                 this.mergeWithNeighbours();
             }
 
@@ -529,7 +_,8 @@
     @Override
     public Entity teleport(TeleportTransition teleportTransition) {
         Entity entity = super.teleport(teleportTransition);
-        if (!this.level().isClientSide && entity instanceof ItemEntity itemEntity) {
+        ItemStack item = this.getItem(); // cookie - short cut count check
+        if (item.getCount() < item.getMaxStackSize() && !this.level().isClientSide && entity instanceof ItemEntity itemEntity) { // cookie - short cut count check
             itemEntity.mergeWithNeighbours();
         }
 
